### PR TITLE
Jhbate/pw logging

### PR DESCRIPTION
**Pull request by jh-bate**
@darinkrauss this should address https://github.com/tidepool-org/blip/issues/248

I did a bit of cleanup also (tried not to go too far) around error handling for creating a user and childUser . 

NOTE this needs to come after https://github.com/tidepool-org/shoreline/pull/19 as I branched from the wrong place

### CODE REVIEW

**Comment by jh-bate** [comment]
@darinkrauss based on my testing yesterday I did a bunch more cleanup in here. I changed to use a logger so that we get the file and line number where we are logging. There were also a few place that logging was missing when it was taking a non 'good' path - i.e. not 200 status OK

```
api/user api.go:173: 409 User already exists
api/user api.go:372: 401 SessionToken: userId not set
api/user api.go:484: 400 Missing id and/or password
api/user api.go:573: 401 Authorization header is invalid
api/user api.go:451: 401 No user matched the given details
api/user api.go:498: 400 Missing id and/or password
```

File: user/api.go [hunk 1]
```diff
@@ -176,197 +206,213 @@ func (a *Api) CreateChildUser(res http.ResponseWriter, req *http.Request) {
 // status: 500 STATUS_ERR_UPDATING_USR
 func (a *Api) UpdateUser(res http.ResponseWriter, req *http.Request, vars map[string]string) {
 
-	if td := getUnpackedToken(req.Header.Get(TP_SESSION_TOKEN), a.ApiConfig.Secret); td != nil {
```
Comment (darinkrauss) [inline user/api.go]: We check `td != nil`, but we do not in the updated code.  Will td ever be nil if err is nil?  Or is checking the err enough?

File: user/api.go [hunk 2]
```diff
@@ -176,197 +206,213 @@ func (a *Api) CreateChildUser(res http.ResponseWriter, req *http.Request) {
...
-					res.WriteHeader(http.StatusOK)
-					return
+					a.logMetric("userupdated", req.Header.Get(TP_SESSION_TOKEN), map[string]string{"server": "false"})
 				}
+				res.WriteHeader(http.StatusOK)
+				return
 			}
 		}
 	}
-	res.WriteHeader(http.StatusUnauthorized)
```
Comment (darinkrauss) [inline user/api.go]: Is this case still handled?  I may have missed it with the other changes?

File: user/api.go [hunk 3]
```diff
@@ -176,197 +206,213 @@ func (a *Api) CreateChildUser(res http.ResponseWriter, req *http.Request) {
...
-	return
+
 }
 
 //Pull the incoming user feilds to search for from http.Request body and
 //find any matches returning them with return http.StatusOK
 func (a *Api) GetUserInfo(res http.ResponseWriter, req *http.Request, vars map[string]string) {
 
-	if td := getUnpackedToken(req.Header.Get(TP_SESSION_TOKEN), a.ApiConfig.Secret); td != nil {
```
Comment (darinkrauss) [inline user/api.go]: Check of td, again?

File: user/token.go [hunk 4]
```diff
@@ -33,74 +32,65 @@ const (
...
-	} else {
-		if err = sessionToken.Save(store); err != nil {
-			log.Print(USER_API_PREFIX, "error saving SessionToken", err.Error())
-			return nil, err
-		} else {
-			return sessionToken, nil
-		}
 	}
+	err = sessionToken.Save(store)
+	return sessionToken, err
```
Comment (darinkrauss) [inline user/token.go]: Probably shouldn't return the sessionToken here is err != nil. So:
```
err = sessionToken.Save(store)
if err != nil {
  return nil, err
}
return sessionToken, nil
```

File: user/api.go [hunk 5]
```diff
@@ -604,15 +643,22 @@ func (a *Api) ServerCheckToken(res http.ResponseWriter, req *http.Request, vars
 		tokenString := vars["token"]
 
 		svrToken := &SessionToken{Id: tokenString}
-		if td := svrToken.UnpackAndVerify(a.ApiConfig.Secret); td != nil && td.Valid {
```
Comment (darinkrauss) [inline user/api.go]: td nil check necessary?

File: user/api.go [hunk 6]
```diff
@@ -176,197 +206,213 @@ func (a *Api) CreateChildUser(res http.ResponseWriter, req *http.Request) {
...
 		}
 	}
-	res.WriteHeader(http.StatusUnauthorized)
-	return
 }
 
 func (a *Api) DeleteUser(res http.ResponseWriter, req *http.Request, vars map[string]string) {
 
-	if td := getUnpackedToken(req.Header.Get(TP_SESSION_TOKEN), a.ApiConfig.Secret); td != nil {
+	td, err := getUnpackedToken(req.Header.Get(TP_SESSION_TOKEN), a.ApiConfig.Secret)
```
Comment (darinkrauss) [inline user/api.go]: td nil check necessary?

**Comment by darinkrauss** [comment]
A few sanity checking comments.  Otherwise, LGTM. :shipit:

File: user/api.go (hunk 1 above)
Comment (jh-bate) [inline user/api.go]: I feel the err check will do now. The function always returns an err if no token is returned

File: user/api.go (hunk 2 above)
Comment (jh-bate) [inline user/api.go]: We do this as part of the token check now - if there is an issue with the token or its not valid then that happens first thing - the actual error is logged also

File: user/token.go (hunk 4 above)
Comment (jh-bate) [inline user/token.go]: yeah good point - will update

**Pull request merged**